presto-main/src/main/java/com/facebook/presto/execution/QueryStateTimer.java

```diff
@@ -62,14 +64,26 @@ public QueryStateTimer(Ticker ticker)
     // State transitions
     //
 
+    public void beginQueued()
+    {
+        beginQueued(tickerNanos());
+    }
+
+    private void beginQueued(long now)
+    {
+        waitingForPrerequisitesTime.compareAndSet(null, nanosSince(createNanos, now));
+        beginQueuedNanos.compareAndSet(null, now);
+    }
+
     public void beginWaitingForResources()
     {
         beginWaitingForResources(tickerNanos());
     }
 
     private void beginWaitingForResources(long now)
     {
-        queuedTime.compareAndSet(null, nanosSince(createNanos, now));
+        beginQueued(now);
+        queuedTime.compareAndSet(null, nanosSince(beginQueuedNanos, now));
```

Comment:
it seems this two statement makes queuedTime almost 0. Is that intentional?

Reply:
This is how this class works in general. Here is another function : 

```
private void beginDispatching(long now)
    {
        beginWaitingForResources(now);
        resourceWaitingTime.compareAndSet(null, nanosSince(beginResourceWaitingNanos, now));
        beginDispatchingNanos.compareAndSet(null, now);
}
```

The logic is as following:

1. `beginQueued(now)` -> `beginQueued` has `compareAndSet` - this is there for situations where in the query goes from `WAITING_FOR_PREREQUISITES` to `FAILED` - in that case this chaining of functions ensures that all the states are set. So if the query had already crossed queuing state - compareAndSet will make this function a no-op
2. `queuedTime.compareAndSet(null, nanosSince(beginQueuedNanos, now));` - the start of WaitingForResources indicates the end of queuing and hence we are setting the queuing time now